### PR TITLE
tests: metacoin

### PR DESCRIPTION
**Pull request by hrls**

### CODE REVIEW

**Comment by vldm** [comment]
I have converted this pullrequest to draft, because it depend on #77

File: evm-utils/evm-state/src/evm_backend.rs [hunk 1]
```diff
@@ -35,6 +37,9 @@ impl EvmBackend {
                 } => {
                     debug!("Apply::Modify address = {}, basic = {:?}", address, basic);
 
+                    let storage = HashMap::<H256, H256>::from_iter(storage);
```
Comment (vldm) [inline evm-utils/evm-state/src/evm_backend.rs:40]: can we hide this behind debug cfg?

File: evm-utils/evm-state/src/lib.rs [hunk 2]
```diff
@@ -288,11 +288,353 @@ mod tests {
...
+                .transaction_execute(create_tx, noop_precompile)
+                .unwrap(),
+            (ExitReason::Succeed(ExitSucceed::Returned), _)
+        ));
+
+        alice.nonce += 1;
+
+        let call_tx = alice.call(
+            contract,
+            &metacoin::GET_BALANCE
```
Comment (vldm) [inline evm-utils/evm-state/src/lib.rs:432]: We can wrap this routine, into highlevel api:

```
let tx = create_call_tx();
executor.exec(tx);
decode_output(bytes);
```
to
```
trait ExecutorTest {
fn call<T: ParseArgs, Out: ParseOutput>(&mut self, args: T) -> Result<Out>
```

**Review by vldm — APPROVED** [review]
1. Can you move metacoin files into tests folder?
2. Lets plan to use solc as dev deps in order to compile code.

File: evm-utils/evm-state/src/evm_backend.rs (hunk 1 above)
Comment (hrls) [reply evm-utils/evm-state/src/evm_backend.rs:40]: Looks like it is impossible without cloning (and thus `Clone` became restriction of apply `I` type) or other type specific for making iterator traversable twice.

**Comment by hrls** [comment]
> * Can you move metacoin files into tests folder?

done
> * Lets plan to use solc as dev deps in order to compile code.

done

File: evm-utils/evm-bridge/src/main.rs [hunk 3]
```diff
@@ -1062,7 +1061,7 @@ fn send_and_confirm_transactions<T: Signers>(
     mut transactions: Vec<solana::Transaction>,
     signer_keys: &T,
 ) -> StdResult<(), Box<dyn std::error::Error>> {
-    const SEND_RETRIES: usize = 5;
```
Comment (vldm) [inline evm-utils/evm-bridge/src/main.rs:1065]: can you return this back)

File: evm-utils/evm-bridge/src/main.rs [hunk 4]
```diff
@@ -1077,15 +1076,19 @@ fn send_and_confirm_transactions<T: Signers>(
                     sleep(Duration::from_millis(1000 / DEFAULT_TICKS_PER_SECOND));
                 }
 
+                debug!("Sending {:?}", transaction.signatures);
+
                 let signature = rpc_client
                     .send_transaction_with_config(
                         &transaction,
                         RpcSendTransactionConfig {
-                            skip_preflight: true,
```
Comment (vldm) [inline evm-utils/evm-bridge/src/main.rs:1084]: Remove note

File: evm-utils/evm-bridge/src/main.rs [hunk 5]
```diff
@@ -1119,93 +1124,147 @@ fn send_and_confirm_transactions<T: Signers>(
...
+        &storage_pubkey,
+        balance,
+        tx_bytes.len() as u64,
+        &solana_evm_loader_program::ID,
+    );
 
-    let signers = [keypair];
-    let create_account_tx = solana::Transaction::new(&signers, message, blockhash);
+    let allocate_storage_ix =
+        solana_evm_loader_program::big_tx_allocate(&storage_pubkey, tx_bytes.len());
```
Comment (vldm) [inline evm-utils/evm-bridge/src/main.rs:1176]: Why you need alocate instruction?

File: evm-utils/evm-bridge/src/main.rs [hunk 6]
```diff
@@ -1119,93 +1124,147 @@ fn send_and_confirm_transactions<T: Signers>(
...
+                "Execute EVM tx at {} was done, signature = {:?}",
+                storage_pubkey, signature
+            )
+        })
+        .map_err(|e| {
+            error!("Execute EVM tx at {} failed: {:?}", storage_pubkey, e);
+            e
+        })?;
+
+    // TODO: here we can transfer back lamports and delete storage
```
Comment (vldm) [inline evm-utils/evm-bridge/src/main.rs:1267]: I am not sure that we should. Let better in Finalize, just move lamports to EvmState account.?

File: evm-utils/evm-state/src/evm_backend.rs (hunk 1 above)
Comment (vldm) [reply evm-utils/evm-state/src/evm_backend.rs:40]: And why you need cloning?

File: evm-utils/evm-state/src/layered_backend.rs [hunk 7]
```diff
@@ -1,446 +1,389 @@
...
+        }
+
+        for (hash, receipt) in std::mem::take(&mut self.receipts) {
+            self.kvs.set::<Receipts>(hash, receipt);
+        }
+
+        self.root = new_root;
+    }
 
+    /// Ignores all unapplied updates.
```
Comment (vldm) [inline evm-utils/evm-state/src/layered_backend.rs:146]: hmm, can you add warn about unapplied updates. Is it possible to have unapplied updates there? Maybe even assert is needed?

File: evm-utils/evm-state/src/layered_backend.rs [hunk 8]
```diff
@@ -1,446 +1,389 @@
...
-pub struct BigTransactionStorage {
-    pub tx_chunks: Vec<u8>,
-}
 
-#[derive(Clone, Debug)] // TODO: Debug
-pub struct EvmState {
-    pub(crate) current_slot: Slot,
-    pub(crate) previous_slot: Option<Slot>,
+impl EvmState {
+    pub fn apply(&mut self) {
```
Comment (vldm) [inline evm-utils/evm-state/src/layered_backend.rs]: rename to commit

**Review by vldm — APPROVED** [review]

File: evm-utils/evm-state/src/evm_backend.rs (hunk 1 above)
Comment (hrls) [reply evm-utils/evm-state/src/evm_backend.rs:40]: Do you know other way to iterate on `impl IntoIterator<Item = _>` twice? First for logging, and then for call extend storage. Ok, we can do `.into_iter().map(|item| { log(item); item })` but these extra lines for each entry are kind of mess.

File: evm-utils/evm-state/src/layered_backend.rs (hunk 8 above)
Comment (hrls) [reply evm-utils/evm-state/src/layered_backend.rs]: Done

**Pull request merged**